dataset/src/main/scala/frameless/TypedDataset.scala

```diff
@@ -276,6 +279,27 @@ object TypedDataset {
     sqlContext: SQLContext
   ): TypedDataset[A] = {
     val dataset = sqlContext.createDataset(data)(TypedExpressionEncoder[A])
-    new TypedDataset[A](dataset)
+    TypedDataset.create[A](dataset)
+  }
+
+  def create[A: TypedEncoder](dataset: Dataset[A]): TypedDataset[A] = {
+    val e = TypedEncoder[A]
+    val output = dataset.queryExecution.analyzed.output
+
+    val fields = TypedExpressionEncoder.targetStructType(e)
+    val colNames = fields.map(_.name)
+
+    val shouldRename = output.zip(colNames).exists {
+      case (expr, colName) => expr.name != colName
+    }
+
+    val renamed =
+      // some invariant has failed, let Spark generate error
+      if (output.size != fields.size) dataset.toDF(colNames: _*)
```

Comment:
The being said if there is such shortciruit, `shouldRename` and it's deps should probably be lazily computer (`def`).